### DISCUSSION TITLE
API Rules for Record Operations in a Go Backend

### DISCUSSION THREAD

**Discussion by Salzstangenesser** [pocketbase/pocketbase#5908] [Q&A] [answered]
Hello everyone,

I have a question about how API rules apply to record operations when working with Pocketbase. I am using Pocketbase prerelease `v0.23.0-rc14` and Go `1.23`.

**Do API rules also apply to record operations performed directly in Go code?**

For context, I have a collection called `invoices` with fields: `ID`, `userID` (linked), and `value`. The API rules for this collection are defined as follows:

- **listRule**: `@request.auth.id = userID.id`
- **viewRule**: `@request.auth.id = userID.id`
- **createRule**: `@request.auth.id != ""`
- **updateRule**: `@request.auth.id = userID.id`
- **deleteRule**: `@request.auth.id = userID.id`

Given these rules, I would expect that only the invoices belonging to the authenticated user are returned. However, the following code returns **all invoices**, regardless of the API rules:

```go
app.OnServe().BindFunc(func(e *core.ServeEvent) error {
    e.Router.GET("/invoices", func(re *core.RequestEvent) error {
        records, err := re.App.FindAllRecords("invoices")
        // No filtering seems to be applied here
    })
    return e.Next()
})
```
This suggests that API rules do not automatically apply to Go backend operations, and the backend code can access all records unless explicit filters are implemented.

Here’s one approach I’ve tried to restrict the results to the authenticated user's invoices:

```go
app.OnServe().BindFunc(func(e *core.ServeEvent) error {
    e.Router.GET("/invoices", func(re *core.RequestEvent) error {
        records, err := re.App.FindRecordsByFilter(
            "invoices", 
            "userID = {:userID}", 
            "", 
            0, 
            0, 
            dbx.Params{"userID": re.Auth.Id},
        )
        // Filter based on authenticated user
    })
    return e.Next()
})
```
This works, but it requires manually adding filters for every query. I was hoping to rely on the API rules for global enforcement, keeping them centralized and avoiding redundant code.

**Questions**:

1. Is it expected behavior that API rules don’t apply to Go backend operations?
2. If yes, is there a recommended way to enforce API rules globally without adding filters to every query?
3. Could this be an implementation error on my part? (I'm relatively new to both Go and Pocketbase.)

I’m also using the prerelease version because I wanted to get familiar with the new system rather than learning the older one. If this behavior is by design, I’d appreciate any advice on best practices to handle this.

Thank you in advance for your help!

**Comment by chnfyi** [comment]
Hi,

Yes it is the expected behavior that API rules doesn't apply to Go backend operations. They perform access control and permissions for API requests. If you want to apply similar process in your Go code, you'll need to implement the logic manually in your hooks.

**Comment by ganigeorgiev** [comment, marked as answer]
As others already mentioned, the list/view/create/update/delete API rules apply only for their designated [API routes](https://pocketbase.io/docs/api-records/#crud-actions).

We don't perform API rules check out of the box for custom routes because we don't know what the custom route is supposed to do and who should have access to it (e.g. you may want to generate a report or some other aggregation based on a some custom query).

Additionally note that when you call `re.App.FindAllRecords`, `re.App.FindRecordsByFilter`, etc. you are fetching data directly from the database and these functions can be used outside of a request context, e.g. as part of a custom console command, cron task, etc. where there is no such thing as "authenticated user" to perform access checks against it.

I don't know what to recommend because it is not clear what is the purpose of the custom routes and why they need to replicate the already existing Web APIs.

**Comment by Salzstangenesser** [comment]
Thank you for clarifying that API rules don’t apply to Go backend operations and that they are designed specifically for designated API routes. I initially assumed that since App was invoked from a request context, it might internally use re.Auth for access control.

My project uses htmx for server-side rendering (SSR), where the server handles the request and returns HTML to the client. Because some queries are reused across different locations, I’ve created a Go package for queries. For example, one of the query functions looks like this:

```go
func GetAllInvoices(re *core.RequestEvent) ([]*core.Record, error) {
    records, err := re.App.FindRecordsByFilter(
        "invoices",
        "userID = {:userID}",
        "",
        0,
        0,
        dbx.Params{"userID": re.Auth.Id},
    )
    if err != nil {
        return nil, err
    }
    return records, nil
}
```
This approach works, but it requires me to manually define filters for every query. It would have been convenient if these queries could automatically respect the API rules, reducing redundancy. However, now I understand that I need to define the rules explicitly in each query and bypass API rules entirely.

I also tried an approach provided by Copilot to call the API directly, leveraging API rules like this:

```go
type Invoices struct {    
    ID     string `json:"id"`
    UserID string `json:"userID"`
    Value  int    `json:"value"`    
}

func GetAllInvoices(userID, authToken string) ([]*Invoices, error) {
    url := fmt.Sprintf("http://your-pocketbase-url/api/collections/invoices?filter=userID='%s'", userID)

    req, err := http.NewRequest("GET", url, nil)
    if err != nil {
        return nil, err
    }

    req.Header.Add("Authorization", "Bearer "+authToken)

    client := &http.Client{}
    resp, err := client.Do(req)
    if err != nil {
        return nil, err
    }
    defer resp.Body.Close()

    if resp.StatusCode != http.StatusOK {
        return nil, fmt.Errorf("failed to fetch records: %s", resp.Status)
    }

    body, err := ioutil.ReadAll(resp.Body)
    if err != nil {
        return nil, err
    }

    var invoices []*Invoices
    err = json.Unmarshal(body, &invoices)
    if err != nil {
        return nil, err
    }

    return invoices, nil
}
```

Using this approach, I could utilize the API rules, but it adds overhead since I’d need to manage API requests and responses for queries that otherwise could be directly handled in Go. I may stick with defining rules directly in my query functions.

Finally, I’d suggest adding the information that API rules don’t apply to framework-level operations to the documentation for better clarity. It would be helpful to include this detail in sections like:

- [Use as Framework](https://pocketbase.io/docs/use-as-framework/)
- [API Rules and Filters](https://pocketbase.io/docs/api-rules-and-filters/)
- [Go Records](https://pocketbase.io/docs/go-records/)

Thanks again for your assistance and for pointing me in the right direction!

**Reply by ganigeorgiev** [reply]
> My project uses htmx for server-side rendering (SSR) ...

I don't understand why so many users are fixated of using htmx, whether is just a hype or they really enjoy working with it,  but I don't recommend using it with PocketBase (see https://github.com/pocketbase/pocketbase/discussions/5307, https://github.com/pocketbase/pocketbase/discussions/4161 and probably others).

PocketBase is designed to be used directly from the client-side (e.g. browser SPA or mobile app). Go SSR is certainly doable but replicating the CRUD routes with all of it caveats around pagination/sorting/filtering, overwriting the auth handling and the realtime events,  dealing with cookies, etc. is just too much work for very little benefit and for me kind of defeat the purpose of using PocketBase in the first place.

> Finally, I’d suggest adding the information that API rules don’t apply to framework-level operations to the documentation for better clarity.

I think it is already clear enough but I will consider phrasing it better.
